DRAFT - FM_PC_ExternalPriceSetter, Oracle, blacklist contracts and interfaces implemented

vendrell46 opened the pull request:
The scope of this PR is to review the proposed implementation, considering that we DO NOT consider it complete, but rather a work in progress.

We want to ensure we are going in the right direction by getting feedback from the team.

vendrell46 committed:
Contract structure defined

vendrell46 committed:
removed IFundingManager extension

vendrell46 committed:
IFundingManager_v1 needed

vendrell46 committed:
extended IRedeemingBondingCurveBase_v1 as well

vendrell46 committed:
ExternalPriceSetter, blacklist, Oracle and interfaces added

Zitzak commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+        uint256 tokenAmount_ = _oracle.getPriceForIssuance() * depositAmount_;
+
+        // Convert mint amount to issuing token decimals
+        mintAmount_ = FM_BC_Tools._convertAmountToRequiredDecimal(
```
I think the conversion has to happen before the calculation right?

vendrell46 replied:
ouch, yes thank you!

Zitzak reviewed (changes requested):
Thank you for the work. Overall looks great and right on track. I've left a comments which are mostly about style, formatting etc and some related to how the implementation might be done differently.

Running a bit out of time on my end to do a second pass but I think this will give you guys enough to keep developing.

vendrell46 commented on src/modules/fundingManager/oracle/ExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,66 @@
+*/
+
+// @audit-info  change name to ExternalPrice 
+contract ExternalPriceSetter_v1 is IOraclePrice_v1, Ownable2StepUpgradeable { 
```
done

vendrell46 commented on src/modules/fundingManager/oracle/ExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,66 @@
+
+/**
+* @title   Oracle Price Implementation for USP Token
+* @notice  Manages and provides daily price data for USP token operations
```
done

vendrell46 commented on src/modules/fundingManager/oracle/ExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,66 @@
+    /// @notice Sets the issuance price
+    /// @param price_ New issuance price
+    function setIssuancePrice(uint256 price_) external onlyOwner {
+        if (price_ == 0) revert InvalidPrice();
```
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -0,0 +1,22 @@
+* @custom:security-contact security@inverter.network
+* @author  Zealynx Security
+*/
+interface IOraclePrice_v1 {
```
When we talked last time you mentioned to keep an oracle interface because eventually this will/might come from an actual oracle, so I thought of leaving it for that purpose.

vendrell46 replied:
formatted

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,132 @@
+ * @custom:security-contact security@inverter.network
+ * @author  Zealynx Security
+ */
+contract ERC20Issuance_blacklist_v1 is 
```
done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,132 @@
+import { ERC20 } from "@oz/token/ERC20/ERC20.sol";
+
+/**
+ * @title   Blacklist-enabled ERC20 Issuance Token
```
done

vendrell46 commented on src/modules/fundingManager/oracle/ExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
+
```
done

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,41 @@
+   /// @param account_ The address to remove
+   function removeFromBlacklist(address account_) external;
+
+   /// @notice Adds multiple addresses to blacklist
```
done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,132 @@
+    // Constants
+
+    /// @dev Maximum number of addresses that can be blacklisted in a batch
+    uint256 public constant BATCH_LIMIT = 50;
```
I'd say, 200 will most likely be a safe amount

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,132 @@
+    //--------------------------------------------------------------------------
+    // Events
+
+    event AddedToBlacklist(address indexed account_);
```
done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,132 @@
+
+    /// @inheritdoc IERC20Issuance_blacklist_v1
+    function addToBlacklist(address account_) public onlyOwner {
+        require(!isBlacklisted(account_));
```
done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,132 @@
+        address[] calldata accounts_
+    ) external onlyOwner {
+        uint256 totalAccount_ = accounts_.length;
+        require(totalAccount_ <= BATCH_LIMIT, "Batch limit exceeded");
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+    mapping(address => bool) private _whitelist;
+
+    /// @dev Storage gap for future upgrades
+    uint256[50] private __gap;
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+    uint256 private _nextOrderId;
+
+    // For tracking total open redemption amount
+    uint256 public openRedemptionAmount;
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+
+    /// @dev Ensures caller is whitelisted
+    modifier onlyWhitelisted() {
+        if (!_whitelist[msg.sender]) {
```
done and done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+
+    /// @notice Adds a single address to the whitelist
+    /// @param account_ The address to whitelist
+    function addToWhitelist(address account_) external onlyOrchestratorAdmin {
```
done

vendrell46 replied:
alright, I understand now what you meant. I've removed the functions, left the role `WHITELISTED_USER` in the contract. The grants and revokes will be done through `AUT_Roles_v1` contract.

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+    }
+
+    // @audit-info Scenario: Display expected token return amount to the buyer
+    function calculateExpectedReturn(
```
no reason, at first I thought it would need to be implemented and then I realize we just need to call another function.

Deleted

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+        uint256 collateralAmount_,
+        uint256 minAmountOut_
+    ) public override(IFM_PC_ExternalPrice_Redeeming_v1, BondingCurveBase_v1) onlyWhitelisted notBlacklisted {
+        buyFor(msg.sender, collateralAmount_, minAmountOut_);  
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+    function transferOrchestratorToken(
+        address to_,
+        uint256 amount_
+    ) external override(IFundingManager_v1) onlyPaymentClient {
```
I thought it was, but it compiles without it, so I will remove it

vendrell46 committed:
formatting files and adapting to the code of conduct

vendrell46 committed:
ManualExternalPriceSetter_v1 modified as LM

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+        _orderId = _nextOrderId++;
+
+        // Create and store order
+        _redemptionOrders[_orderId] = RedemptionOrder({
```
struct removed and we've introduced `_addPaymentOrder()`

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+
+    /// @notice Sets fee for sell operations
+    /// @param fee_ New fee amount
+    function setSellFee(
```
max fee check added

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,132 @@
+    /// @notice Mints tokens if account is not blacklisted
+    /// @param account_ Address to mint to
+    /// @param amount_ Amount to mint
+    function mintAllowed(address account_, uint256 amount_) public onlyMinter {
```
`_update()` and restriction added

vendrell46 committed:
added fees checks, paymentOrder to the process and defined role for blacklist

Zitzak commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+        public
+        override(IFM_PC_ExternalPrice_Redeeming_v1, BondingCurveBase_v1)
+        onlyModuleRole(WHITELISTED_ROLE)
+        notBlacklisted
```
This one can be removed as the ERC20 blacklist reverts automatically during transfers if an address is blacklisted

vendrell46 replied:
removed

Zitzak commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+    // Public Functions
+
+    /// @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    /// @param collateralAmount_ The amount of collateral to spend
```
Please add a dev tag explaining the whitelisted restricted

Zitzak replied:
One other thing to note, we want to restrict the buyFor and sellTo functions as the client doesn't want user to redeem to other addresses or buy for someone else. How should we approach this such that it can be toggled on/off?

vendrell46 replied:
handled as proposed on slack

Zitzak commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,177 @@
+
+    /// @notice Buys tokens with provided collateral
+    /// @param collateralAmount_ The amount of collateral to spend
+    function buy(uint collateralAmount_, uint minAmountOut_) external;
```
I don't think we need to define the buy and sell function in the interface as they are defined in the base contracts

vendrell46 replied:
removed from interface

Zitzak reviewed (changes requested):
Thank you for the work guys!

I have reviewed everything again until the sell function in the FM. I will need to come back to finish the review tomorrow

FHieser commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,132 @@
+    // External Functions
+
+    /// @inheritdoc IERC20Issuance_blacklist_v1
+    function addToBlacklist(address account_) public onlyOwner {
```
Adding it to an additional role definitely gives it more flexibility. Sure lets do that

vendrell46 replied:
done

FHieser commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+
+    /// @param interfaceId_ The interface identifier to check support for
+    /// @return True if the interface is supported
+    function supportsInterface(
```
Nope :P
It needs to check for the IFM_PC_ExternalPrice_Redeeming_v1 interface instead of IRedeemingBondingCurveBase_v1.
Because of the inheritance of the ERC20PaymentClientBase_v1 and RedeemingBondingCurveBase_v1 contracts you already check for their respective Interfaces, so IRedeemingBondingCurveBase_v1 for example is already done

FHieser commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+
+    /// @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    /// @param collateralAmount_ The amount of collateral to spend
+    function buy(
```
yeah i agree here, doesnt need to be included in the interface 
Did you try building it?
Just curious if this causes alerts in some way

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+
+// Internal
+import {IFM_PC_ExternalPrice_Redeeming_v1} from
+    "./interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol";
```
done

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
+
```
done

vendrell46 commented on src/modules/fundingManager/oracle/ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,102 @@
+ * @author  Zealynx Security
+ */
+
+contract ManualExternalPriceSetter_v1 is IManualExternalPriceSetter_v1, Module_v1, IOraclePrice_v1 {
```
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,53 @@
+    // External Functions
+
+    /// @notice Sets the issuance price
+    /// @param price_ New issuance price (must be non-zero)
```
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,53 @@
+    // Errors
+
+    /// @notice Thrown when attempting to set a price to zero
+    error ExternalPriceSetter__InvalidPrice();
```
done

vendrell46 commented on src/modules/fundingManager/oracle/ExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /// @inheritdoc IOraclePrice_v1
+    function getPriceForIssuance() external view returns (uint256) {
```
sorry missed this comment.

we will be normalizing the prices to deal with them internally when needed.

Please, let me know what you think.

Zitzak replied:
Leaving it open for now. I left another comment/clarification question around this

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_Oracle_Redeeming_v1.sol:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @author  Zealynx Security
+ */
+interface IFM_PC_Oracle_Redeeming_v1 is
```
true. removed

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,173 @@
+    //--------------------------------------------------------------------------
+    // Modifiers
+
+    modifier onlyAllowed(address account_) {
```
good eye! removed

vendrell46 committed:
managing roles, renaming, and removing unneeded things

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,107 @@
+    // Errors
+
+    /// @notice Thrown when attempting to blacklist the zero address
+    error ERC20Issuance__ZeroAddress();
```
done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,173 @@
+
+    /// @inheritdoc IERC20Issuance_Blacklist_v1
+    function addToBlacklist(address account_) public onlyOwner {
+        if (isBlacklisted(account_)) {
```
thanks for pointing it out. changed

Zitzak replied:
Thank you. Could you also change it the same for removing an address from the blacklist?

vendrell46 replied:
done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,173 @@
+ *
+ * @author  Zealynx Security
+ */
+contract ERC20Issuance_Blacklist_v1 is
```
done

Zitzak replied:
Yes you are right, like the `isBlacklisted` function

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,173 @@
+        override
+    {
+        if (isBlacklisted(from) || isBlacklisted(to)) {
+            revert ERC20Issuance__BlacklistedAddress(from);
```
ahhg, you're right. Changed.

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,173 @@
+    function _update(address from, address to, uint amount)
+        internal
+        virtual
+        override
```
I've added `override(ERC20Capped)` to be more specific

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,173 @@
+    /// @notice Mints tokens if account is not blacklisted
+    /// @param account_ Address to mint to
+    /// @param amount_ Amount to mint
+    function mintAllowed(address account_, uint amount_) public onlyMinter {
```
makes sense

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,173 @@
+    /// @notice Redeems tokens if account is not blacklisted
+    /// @param account_ Address to redeem from
+    /// @param amount_ Amount to redeem
+    function redeem(address account_, uint amount_)
```
done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,173 @@
+    {
+        // To be implemented
+    }
+
```
I've added 
`    //--------------------------------------------------------------------------`

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+    // Constants
+
+    /// @dev Maximum number of addresses that can be whitelisted in a batch
+    uint private constant MAX_BATCH_SIZE = 50;
```
I've removed the addBatch whitelist function, as proposed in another comment. So, I've also removed this constant now.

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+    uint8 private constant EIGHTEEN_DECIMALS = 18;
+
+    /// @notice Role for whitelisted addresses
+    bytes32 public constant WHITELISTED_ROLE = "WHITELISTED_USER";
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+    /// @dev Token decimals of the Orchestrator token
+    uint8 private _collateralTokenDecimals;
+
+    // For order IDs
```
natspec added

vendrell46 committed:
whitelist logic removed, blacklist fixes

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+
+    /// @notice Maximum fee that can be charged for sell operations, expressed in basis points (10000 = 100%)
+    /// @dev Set to 1000 basis points (10%)
+    uint256 private constant MAX_SELL_FEE = 1000; // 10% in basis points
```
I've modified all this and also introduced max fee setters & getters restricted to `onlyOrchestratorAdmin`. Let me know if that works.

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+
+    /// @notice Removes an address from the whitelist
+    /// @param account_ The address to remove from whitelist
+    function removeFromWhitelist(address account_)
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+    /// @notice Checks if an address is whitelisted
+    /// @param account_ The address to check
+    /// @return isWhitelisted_ True if address is whitelisted
+    function isWhitelisted(address account_)
```
yup. It's removed now

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+
+    /// @param interfaceId_ The interface identifier to check support for
+    /// @return True if the interface is supported
+    function supportsInterface(bytes4 interfaceId_)
```
moved

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+    }
+
+    /// @notice Gets current static price for buying tokens
+    /// @return price_ The current buy price
```
done

Zitzak replied:
Not actually done :D

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+    //--------------------------------------------------------------------------
+    // External Functions
+
+    // @audit-info Scenario: Project deposits collateral into the FM
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
+
+//--------------------------------------------------------------------------
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,495 @@
+    // @audit-info Scenario: Project deposits collateral into the FM
+    /// @notice Allows depositing collateral to provide reserves for redemptions
+    /// @param amount_ The amount of collateral to deposit
+    function depositReserve(uint amount_) external onlyOrchestratorAdmin {
```
added to interface

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+        if (amount_ == 0) revert FM_ExternalPrice__InvalidAmount();
+        
+        // Transfer collateral from sender to FM
+        IERC20(token()).safeTransferFrom(msg.sender, address(this), amount_);
```
fixed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,455 @@
+
+    /// @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    /// @param collateralAmount_ The amount of collateral to spend
+    function buy(
```
could you please elaborate @FHieser?

What are we missing? 
What should be implemented in this contract for those functions?

vendrell46 replied:
done

Zitzak commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -64,61 +71,69 @@ contract ERC20Issuance_Blacklist_v1 is
 
+    // Modifier to check for blacklist manager role
+    modifier onlyBlacklistManager() {
+        require(hasRole(BLACKLIST_MANAGER_ROLE, msg.sender), "Caller is not a blacklist manager");
```
Please add a custom error for this modifier ser

vendrell46 replied:
done

Zitzak commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -1,6 +1,10 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 pragma solidity 0.8.23;
 
+//--------------------------------------------------------------------------
```
```suggestion
```

Zitzak replied:
I think I copy pasted the the lines here, which shouldn't have been. Apologies for the confusion

vendrell46 committed:
missing logic implemented

vendrell46 committed:
blacklist fixes and custom error added

vendrell46 committed:
switched from role to mapping for blacklist manager

Zitzak commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,181 @@
+        );
+
+        // Store token decimals for price normalization
+        _inputTokenDecimals = IERC20Metadata(inputToken).decimals();
```
Could you lead me through what is meant by input/output token and how this should be contextualized.

I see that it is a solution for the conversion but I'm not sure how it should be used. I think this also feeds into the questions of in which token is the return price for issuance/redemption denominated

vendrell46 replied:
I see, maybe the naming was not ideal.

I've changed them to: 
```
    /// @notice Decimals of the collateral token (e.g., USDC with 6 decimals)
    /// @dev This is the token used to pay/buy with
    uint8 private _collateralTokenDecimals; // previously inputTokenDecimals

    /// @notice Decimals of the issuance token (e.g., ISS with 18 decimals)
    /// @dev This is the token being bought/sold
    uint8 private _issuanceTokenDecimals; // previously outputTokenDecimals
```

Zitzak commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,181 @@
+
+    //--------------------------------------------------------------------------
+    // Interface Support
+
```
Interface can be inherited from ERC165Upgradeable, see [link](https://github.com/InverterNetwork/contracts/blob/main/src/modules/fundingManager/bondingCurve/FM_BC_Bancor_Redeeming_VirtualSupply_v1.sol#L74)

vendrell46 replied:
added

Zitzak commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,181 @@
+    //--------------------------------------------------------------------------
+    // Constants
+
+    bytes32 public constant PRICE_SETTER_ROLE = "PRICE_SETTER_ROLE";
```
Missing natspec for both contants

vendrell46 replied:
added

Zitzak commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,181 @@
+        
+        // Normalize price to internal decimal precision
+        _issuancePrice = _normalizePrice(price_, _inputTokenDecimals);
+        emit PriceSet(price_);
```
We prob only need one of the two emits. Not sure if I commented on that already, making you implement it in this way

vendrell46 replied:
removed `PriceSet`

Zitzak commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,56 @@
+
+    /// @notice Emitted when a new issuance price is set
+    /// @param price The new issuance price
+    event IssuancePriceSet(uint price);
```
Both this event and the RedemptionPriceSet should be emitting indexed price

vendrell46 replied:
done

Zitzak commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,56 @@
+
+    /// @notice Emitted when a new price is set
+    /// @param price The new price that was set
+    event PriceSet(uint256 indexed price);
```
I believe this could be removed. We would be duplicating the events currently otherwise

vendrell46 replied:
removed

Zitzak commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,56 @@
+ */
+interface ILM_ManualExternalPriceSetter_v1 {
+    //--------------------------------------------------------------------------
+    //==========================================================================
```
```suggestion
```

vendrell46 replied:
removed

Zitzak commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -0,0 +1,40 @@
+    // External Functions
+
+    /// @notice Gets current price for token issuance
+    /// @return price_ Current price for buying tokens
```
We might want to add information about in which token the price is denominated here and with the redemption one as well. This would help future devs to implement it knowing what to return. Wdyt?

Zitzak replied:
Could maybe also be added in the interface natspec

vendrell46 replied:
I hope now is clearer. I've added extra info

Zitzak commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author  Zealynx Security
+ */
+interface ILM_ManualExternalPriceSetter_v1 {
```
Not sure if I missed this, but I believe the plan was that this interface would inherit from the IOracle interface. At least that is shown on the Mermaid diagram and you wrote this in the natspec of the interface.

vendrell46 replied:
I had it added in the contract `LM_ManualExternalPriceSetter_v1`. Moved it here.

Zitzak reviewed (changes requested):
Another review. Think we are almost there except for some work on the sell function. We should plan a quick sync around this maybe tomorrow

FHieser reviewed (commented):
All my comments up until this point have been addressed

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,102 @@
+ *
+ * @author  Zealynx Security
+ */
+interface IERC20Issuance_Blacklist_v1 is IERC20Issuance_v1 {
```
moved both

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,183 @@
+    //--------------------------------------------------------------------------
+    // External Functions
+
+    modifier onlyBlacklistManager() {
```
added

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,183 @@
+    function isBlacklisted(address account_) public view returns (bool) {
+        return _blacklist[account_];
+    }
+
```
added

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,102 @@
+    error ERC20Issuance_Blacklist_ZeroAddress();
+
+    /// @notice Thrown when attempting to unblacklist an address that is not blacklisted
+    error ERC20Issuance_Blacklist_NotBlacklisted();
```
removed

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,102 @@
+    error ERC20Issuance_Blacklist_BlacklistedAddress(address account);
+
+    /// @notice Thrown when attempting to blacklist an address that is already blacklisted
+    error ERC20Issuance_Blacklist_AddressAlreadyBlacklisted(address account);
```
removed

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,102 @@
+
+    /// @notice Removes an address from blacklist
+    /// @param account_ The address to remove
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress or ERC20Issuance_Blacklist_NotBlacklisted
```
done

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,102 @@
+
+    /// @notice Removes multiple addresses from blacklist
+    /// @param accounts_ Array of addresses to remove
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress or ERC20Issuance_Blacklist_NotBlacklisted
```
done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,183 @@
+ *          Key features:
+ *              - Individual address blacklisting
+ *              - Batch blacklisting operations
+ *              - Owner-controlled blacklist management
```
seems good, thanks. Added

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,183 @@
+    //--------------------------------------------------------------------------
+    // Storage
+
+    /// @dev Mapping of blacklisted addresses
```
changed

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,183 @@
+            symbol_,
+            decimals_,
+            initialSupply_,
+            initialAdmin_
```
makes sense, done

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,183 @@
+
+    //--------------------------------------------------------------------------
+    // Internal Functions
+
```
added

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,183 @@
+        super._update(from, to, amount);
+    }
+
+    function _setBlacklistManager(address account_, bool privileges_)
```
added

vendrell46 commented on src/modules/fundingManager/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @author  Zealynx Security
+ */
+contract ERC20Issuance_Blacklist_v1 is
```
moved

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
```
changed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
```
changed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
```
changed

vendrell46 commented on src/modules/fundingManager/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
```
changed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+ *              - IRedeemingBondingCurveBase_v1
+ *          Key operations:
+ *              - Token buying with collateral
+ *              - Token selling through redemption queue
```
changed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+
+    // Enum for redemption order states
+    enum RedemptionState {
+        PENDING,
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    // Struct to store order details
+    struct RedemptionOrder {
```
we are using it in `sell` function from `FM_PC_ExternalPrice_Redeeming_v1`

vendrell46 replied:
you're very right, my bad. removed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+
+    /// @notice Emitted when an address is added to the whitelist
+    /// @param account_ The whitelisted address
+    event AddressWhitelisted(address indexed account_);
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+
+    /// @notice Emitted when an address is removed from the whitelist
+    /// @param account_ The removed address
+    event AddressRemovedFromWhitelist(address indexed account_);
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+    /// @notice Emitted when multiple addresses are added/removed from whitelist
+    /// @param accounts_ Array of affected addresses
+    /// @param isAdd_ True if adding to whitelist, false if removing
+    event BatchWhitelistUpdated(address[] accounts_, bool isAdd_);
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+    // @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    /// @notice Allows depositing collateral to provide reserves for redemptions
+    /// @param amount_ The amount of collateral to deposit
+    function depositReserve(uint amount_) external override onlyOrchestratorAdmin {
```
modified

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+    /// @notice Emitted when a redemption order's state changes
+    /// @param orderId Order identifier
+    /// @param newState New state of the order
+    event RedemptionOrderStateUpdated(
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+    // Errors
+
+    /// @notice Thrown when an invalid amount is provided
+    error Module__InvalidAmount();
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+    RedeemingBondingCurveBase_v1
+{
+    /// @param interfaceId_ The interface identifier to check support for
+    /// @return True if the interface is supported
```
added

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+    // View Functions
+
+
+    /// @notice Gets the current open redemption amount
```
changed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+    /// @notice Keeps track of the next available order ID to ensure uniqueness
+    uint private _nextOrderId;
+
+    /// @dev Total amount of tokens currently in redemption process
```
changed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+    uint private _maxSellFee;
+
+    /// @dev Maximum fee that can be charged for buy operations, in basis points
+    /// @notice Maximum allowed fee percentage for buying tokens
```
changed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+        // Emit event with all order details
+        emit RedemptionOrderCreated(
+            _orderId,
+            _msgSender(),
```
hey @Zitzak, I don't understand what you meant here

vendrell46 replied:
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,155 @@
+
+    // Struct to store order details
+    struct RedemptionOrder {
+        address seller;
```
added

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+    /// @param fee_ New fee amount
+    function setSellFee(uint fee_)
+        public
+        override(RedeemingBondingCurveBase_v1, IRedeemingBondingCurveBase_v1)
```
It won't compile without it, that's why it is there

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+    /// @notice Helper function to generate a role ID for this module
+    /// @param role The role to generate an ID for
+    /// @return bytes32 The generated role ID
+    function generateRoleId(bytes32 role) internal view returns (bytes32) {
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+            cliff: 0,
+            end: block.timestamp
+        });
+        _addPaymentOrder(order);
```
thanks for pointing out that example. I've added it now

vendrell46 committed:
moved blacklist files to external, modified sell, and more context added

Zitzak commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,546 @@
+ *
+ * @author  Zealynx Security
+ */
+contract FM_PC_ExternalPrice_Redeeming_v1 is
```
Example from the call
```
    function executeRedemptionQueue() onlyModuleRole(PAYMENT_QUEUE_ROLE){
        __Module_orchestrator.paymentProcessor().executeRedemptionQueue();
    }
```

Zitzak commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,57 @@
+
+    /// @notice Emitted when an issuance price is set
+    /// @param  price The new price that was set
+    event IssuancePriceSet(uint indexed price);
```
We also need to emit the timestamp of when the price has been updated for both issuance and redemption

Zitzak commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,202 @@
+import {ERC20Capped} from "@oz/token/ERC20/extensions/ERC20Capped.sol";
+
+/**
+ * @title   ERC20 Issuance Token with Blacklist Functionality
```
Please add the contract version and Inverter standard version to all of the contract and interface natspecs

See link https://github.com/InverterNetwork/standard?tab=readme-ov-file#file-structure

vendrell46 committed:
PAYMENT_QUEUE_ROLE added, _sellOrder overriden

scab24 committed:
emit_timestamp(IssuancePriceSet,RedemptionPriceSet),NatSpec

Zitzak commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,202 @@
+    mapping(address => bool) private _blacklist;
+
+    /// @notice Mapping of blacklist manager addresses
+    mapping (address => bool) private _isBlacklistManager;
```
Just seeing that we don't have any setter for this role other than the internal one called during init.

I think we need an external function here which is restricted to the owner to set a new manager. You can follow the same patterns as [here](https://github.com/InverterNetwork/contracts/blob/main/src/external/token/ERC20Issuance_v1.sol#L68)

Might be good idea to do the address zero input validation on line `84` in the internal function such that it will also be checked when the public function is called

Zitzak commented on src/external/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,103 @@
+ *                          In case of any concerns or findings, please refer to
+ *                          our Security Policy at security.inverter.network or
+ *                          email us directly!
+ *
```
Missing the versioning in this contract

vendrell46 replied:
done

Zitzak commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,672 @@
+    bytes32 public constant WHITELIST_ROLE = "WHITELIST_ROLE";
+
+    /// @notice Role for payment queue
+    bytes32 public constant PAYMENT_QUEUE_ROLE = "PAYMENT_QUEUE_ROLE";
```
I know that I've typed the name of this role while we were in the call :D but I think it's not descriptive enough. Reason being that payment could also be considered in the context of buying. Could you please rename it to something more descriptive ser.

Maybe `"REDEMPTION_QUEUE_EXECUTOR_ROLE"`  (although this might be a bit long). Or maybe `"QUEUE_MANAGER_ROLE"` 🤔

I leave it up to you

vendrell46 replied:
I chose `QUEUE_MANAGER_ROLE`

Zitzak commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,147 @@
+
+    /// @notice Emitted when a new redemption order is created
+    /// @param orderId Order identifier
+    /// @param seller Address selling tokens
```
```suggestion
    /// @param seller Address selling tokens
    /// @param receiver Address who receives the redeemed tokens
```

vendrell46 replied:
done

Zitzak reviewed (changes requested):
I like the solution you guys found to add the emit to the `_sellOrder` function. However I think we need to adapt it just a bit, as well as the fee collection.
Also I believe the question around the decimals for price setting is still open.

Other than that it looks great, and only added a few small comments

Zitzak committed:
fix: remove collateral transfer & collateral check

- The function `handleCollateralTokensAfterSell` was still expecting to
transfer collateral, which would revert in the current implementation.
- The _sellOrder` function had a conditional check which would let
the contract revert given there will not be any collateral in the
contract

Zitzak committed:
style: run make fmt for formating

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,202 @@
+ *
+ * @author  Zealynx Security
+ */
+contract ERC20Issuance_Blacklist_v1 is
```
modified

Zitzak replied:
Name has not been changed yet ser

vendrell46 replied:
weird, it didn't detect it as a change to commit. I've done that manually through Github. Done

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,202 @@
+ *                          In case of any concerns or findings, please refer to
+ *                          our Security Policy at security.inverter.network or
+ *                          email us directly!
+ *
```
added

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,202 @@
+            emit AddedToBlacklist(account_);
+        }
+    }
+
```
added

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,672 @@
+        );
+    }
+
+    function _sellOrder(
```
moved

vendrell46 committed:
small fixes and documentation added

vendrell46 committed:
docs added for issuance and redemption setters

Zitzak commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        return _token;
+    }
+
+    /// @notice Calculates and returns the static price for buying the issuance token.
```
```suggestion
    /// @inheritdoc IBondingCurveBase_v1
```

Zitzak replied:
I closed the other comment, but this is what I meant by inheriting the natspec se

vendrell46 replied:
oookay, I see now. Thank you for clarifying

Zitzak commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        return _oracle.getPriceForIssuance();
+    }
+
+    /// @notice Calculates and returns the static price for selling the issuance token.
```
```suggestion
    /// @inheritdoc IRedeemingBondingCurveBase_v1
```

vendrell46 replied:
done

Zitzak commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -160,13 +155,23 @@ contract ERC20Issuance_Blacklist_v1 is
+    /// @notice Sets a new blacklist manager
+    /// @dev Only callable by the owner
+    /// @param manager_ Address to set as blacklist manager
+    /// @param allowed_ Whether to grant or revoke blacklist manager role
```
Natspec can be moved to the interface

vendrell46 replied:
done and inherited

Zitzak commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -35,10 +35,10 @@ interface IOraclePrice_v1 {
     /// @notice Gets current price for token issuance (buying tokens)
     /// @return price_ Current price in 18 decimals (collateral tokens per 1 issuance token)
```
This comment might not be accurate, as we actually denormalize the decimals before returning them. Same for the `getPriceForRedemption()` function

vendrell46 replied:
corrected

Zitzak reviewed (changes requested):
I leave the last few comments and will be requesting a review of @marvinkruse today

Zitzak commented:
@vendrell46 additionally I noticed that there are several natspec comments in the FM which could be moved to the interface as well

marvinkruse commented on src/external/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,118 @@
+import {IERC20Issuance_v1} from "@ex/token/IERC20Issuance_v1.sol";
+
+/**
+ * @title   ERC20 Issuance Token with Blacklist Interface
```
This whole NatSpec section of an interface should be a 1-to-1 copy of the one from its contract - so we do not add the word "Interface" here.

vendrell46 replied:
fixed

marvinkruse commented on src/external/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,118 @@
+    // Events
+
+    /// @notice Emitted when an address is added to the blacklist
+    /// @param account_ The address that was blacklisted
```
Small indentation issue (we always press tab after the keyword).
```suggestion
    /// @param  account_ The address that was blacklisted
```

vendrell46 replied:
done

marvinkruse commented on src/external/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,118 @@
+    event AddedToBlacklist(address indexed account_);
+
+    /// @notice Emitted when an address is removed from the blacklist
+    /// @param account_ The address that was removed from blacklist
```
Small indentation issue (we always press tab after the keyword). Also added a word.
```suggestion
    /// @param  account_ The address that was removed from the blacklist
```

vendrell46 replied:
done

marvinkruse commented on src/external/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,118 @@
+    event RemovedFromBlacklist(address indexed account_);
+
+    /// @notice Emitted when a blacklist manager role is granted or revoked
+    /// @param account_ The address that was granted or revoked the role
+    /// @param allowed_ Whether the role was granted (true) or revoked (false)
```
Small indentation issue (we always press tab after the keyword).
```suggestion
    /// @param  account_ The address that was granted or revoked the role
    /// @param  allowed_ Whether the role was granted (true) or revoked (false)
```

vendrell46 replied:
done

marvinkruse commented on src/external/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,118 @@
+    /// @notice Thrown when attempting to blacklist the zero address
+    error ERC20Issuance_Blacklist_ZeroAddress();
+
+    /// @notice Thrown when caller does not have blacklist manager role
```
```suggestion
    /// @notice Thrown when the caller does not have the blacklist manager role
```

vendrell46 replied:
done

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
+
+// Imports
```
```suggestion
```

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    RedeemingBondingCurveBase_v1
+{
+    /// @inheritdoc ERC165Upgradeable
+    /// @param interfaceId_ The interface identifier to check support for
+    /// @return True if the interface is supported
```
The NatSpec is already inherited, doesn't need to be overwritten.
```suggestion
```

vendrell46 replied:
removed

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    //--------------------------------------------------------------------------
+    // State Variables
+
+    /// @dev Oracle price feed contract used for price discovery
```
Order should be: `@notice`, then `@dev` (needs to be applied to anything in this section)

vendrell46 replied:
fixed

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+
+    /// @dev Maximum fee that can be charged for sell operations, in basis points
+    /// @notice Maximum allowed project fee percentage for selling tokens
+    uint private _maxSellFee;
```
General question: I find it a bit confusing to just call it "sellFee", as that could either mean the project or collateral fee - or both. If this is just about the project fee, I'd rename this to clearly indicate this.

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    /// @notice Tracks the sum of all pending redemption orders
+    uint private _openRedemptionAmount;
+
+    /// @dev Flag indicating if direct operations are only allowed
```
Is missing a `@notice` tag I think?

vendrell46 replied:
replaced

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        // Initialize base functionality (should handle token settings)
+        _setIssuanceToken(issuanceToken_);
+
+        // Set fees (checking max fees)
```
The actual settings happens a few lines later, so this is just checking (regarding the comment).

vendrell46 replied:
rephrased

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    //--------------------------------------------------------------------------
+    // Modifiers
+
+    /// @notice Modifier to check if direct operations are only allowed
```
```suggestion
    /// @notice Modifier to check if only direct operations are allowed.
```

vendrell46 replied:
modified

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    /// @inheritdoc IFundingManager_v1
+    /// @dev Only payment clients can transfer orchestrator tokens
+    /// @param to_ The recipient address
+    /// @param amount_ The amount to transfer
```
The inherited NatSpec already explains this, so this is just double and can be removed.

vendrell46 replied:
removed

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    /// @notice Sets fee for sell operations
+    /// @dev Only orchestrator admin can call
```
```suggestion
    /// @dev Only the orchestrator admin can call this function.
```

vendrell46 replied:
modified

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        emit TransferOrchestratorToken(to_, amount_);
+    }
+
+    /// @notice Sets fee for sell operations
```
```suggestion
    /// @notice Sets the fee for sell operations.
```

vendrell46 replied:
done

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        return sellFee_;
+    }
+
+    /// @notice Sets fee for buy operations
```
```suggestion
    /// @notice Sets the fee for buy operations.
```

vendrell46 replied:
done

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    /// @notice Sets fee for buy operations
+    /// @dev Only orchestrator admin can call
```
```suggestion
    /// @dev Only the orchestrator admin can call this function.
```

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    /// @notice Gets the maximum fee that can be charged for buy operations
+    /// @return maxBuyFee_ The maximum fee percentage
```
```suggestion
    /// @return maxBuyFee_ The maximum buy fee
```

vendrell46 replied:
done

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    /// @notice Gets the maximum fee that can be charged for sell operations
+    /// @return maxSellFee_ The maximum fee percentage
```
```suggestion
    /// @return maxSellFee_ The maximum sell fee
```

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    /// @notice Sets the direct operations only flag
+    /// @dev Only orchestrator admin can call
```
```suggestion
    /// @dev Only the orchestrator admin can call this function.
```

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        return _maxSellFee;
+    }
+
+    /// @notice Sets the direct operations only flag
```
Could be bit more like "what is actually happening here" on a general level, at least for the `notice` tag, like:
```suggestion
    /// @notice Toggles whether the contract only allows direct operations or not.
```

vendrell46 replied:
done

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+
+    /// @notice Gets the direct operations only flag
+    /// @return isDirectOperationsOnly_ The current value of the flag
+    function isDirectOperationsOnly() public view returns (bool) {
```
The return value is not named in the function so far.

vendrell46 replied:
modified

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    // Internal Functions
+
+    /// @notice Creates and emits a new redemption order
+    /// @dev    This function wraps the `_createAndEmitOrder` internal function with specified parameters to handle
```
The line is too long, needs a break at 80 characters max.

vendrell46 replied:
fixed

marvinkruse reviewed (changes requested):
Hey, leaving some comments as I pause my review for today before I continue tomorrow morning (CET time)! :) Really great work overall, really appreciate this! 👏 Definitely on the right track, just a few small things to iron out, mainly on the Inverter Standard side. It's not easy, even for us, to fully follow it yet, as it's quite a lot of rules and some of them are quite cryptic, so please don't see my comments or the amount of comments as a bad thing, I like your work! 🙏

Some general comments:
* We end each natspec comment with a dot, like
  * `@notice Maximum number of addresses that can be blacklisted in a batch.`
* Section delimiters should have a whitespace between the `//` and the `---` part
* We prefer to press tab after each NatSpec keyword. I left comments for this here and there, so you probably see what I mean?
* There are cases of `@inheritdoc` tags with just two `//` in front instead of three `///`, found these in `FM_PC_ExternalPrice_Redeeming_v1.sol` specifically.
* Some lines are >80 characters (also including whitespaces at the front, etc.), which need a few line breaks here and there, as we do not want to go above that limit of 80.

Thank you in advance 🙏 Let me know if you have any question or need clarifications or feedback on anything.

vendrell46 committed:
Rename ERC20Issuance_blacklist_v1.sol to ERC20Issuance_Blacklist_v1.sol

vendrell46 committed:
Rename IERC20Issuance_blacklist_v1.sol to IERC20Issuance_Blacklist_v1.sol

vendrell46 committed:
fixed executeRedemptionQueue, docs and renaming

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+pragma solidity 0.8.23;
+
+// Imports
+
```
could we double-check with @Zitzak ? It was suggested by him

Zitzak replied:
We can go with what @marvinkruse suggests, as he is up to date with the Inverter standard.

Apologies for letting you add it in @vendrell46 🙇‍♂️

vendrell46 replied:
removed Imports header

vendrell46 committed:
formatting natspec

Zitzak committed:
feat: add treasury collateral withdraw

scab24 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+ *
+ * @custom:version   1.0.0
+ *
+ * @custom:standard-version  1.0.0
```
I understand, so which format would we go with, the one you suggested in the previous comment or the current one? 
Should we leave 1 or 2 spaces?

vendrell46 replied:
Indeed, using the tabs adds just one space. Fixed

Zitzak committed:
Fix fmt

Zitzak committed:
fix: typo

Zitzak committed:
fix: fmt

Zitzak committed:
Merge pull request #2 from ZealynxSecurity/add_collateral_transfer

Feat: Add collateral transfer to treasury

Zitzak committed:
fix: remove oracle from init

0xNuggan commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,742 @@
+    /// @notice Sets the direct operations only flag.
+    /// @dev    Only orchestrator admin can call.
+    /// @param isDirectOperationsOnly_ The new value for the flag.
+    function setIsDirectOperationsOnly(bool isDirectOperationsOnly_)
```
These two functions are missing from the interface

vendrell46 replied:
As discussed with @Zitzak, this was added because this client needs to have this restriction. That's why from my point of view we should not add it to the interface.

0xNuggan commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,742 @@
+    /// @notice Modifier to check if direct operations are only allowed
+    modifier onlyDirectOperations() {
+        if (_isDirectOperationsOnly) {
+            revert
```
Since this is only used to make buyFor and sellTo revert, I think renaming this to sth like "ThirdPartyOperationsEnabled" would make it clearer when reading the code

vendrell46 replied:
modifier moved

0xNuggan replied:
Thanks! 
Again, I'd consider renaming the modifier. As somebody looking into the code and making sense of it, the function header `buyFor() onlyModuleRole(x) onlyDirectOperations` reads as "the buyFor function can only be called by role X and and only when the directOperations flag is true" , while it actually does the opposite.

vendrell46 replied:
I see your point, renamed to `ThirdPartyOperationsEnabled`

Zitzak committed:
Merge pull request #3 from ZealynxSecurity/removeOracleAddress

fix: remove oracle from init

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+ *              - Batch blacklisting operations
+ *              - Owner-controlled manager role assignment
+ *              - Blacklist manager controlled blacklist management
+ *          All blacklist operations can only be performed by the contract owner.
```
clarified

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+ *                          our Security Policy at security.inverter.network or
+ *                          email us directly!
+ *
+ * @custom:version   1.0.0
```
true, eagle eye. Fixed

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+    // Storage
+
+    /// @notice Mapping of blacklisted addresses
+    mapping(address => bool) private _blacklist;
```
done

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+    mapping(address => bool) private _blacklist;
+
+    /// @notice Mapping of blacklist manager addresses
+    mapping(address => bool) private _isBlacklistManager;
```
done

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+    /// @param initialSupply_ Initial token supply
+    /// @param initialAdmin_ Initial admin address
+    /// @param initialBlacklistManager_ Initial blacklist manager (typically an EOA)
+    constructor(
```
added

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+    // External Functions
+
+    /// @notice Modifier to check if the caller is a blacklist manager
+    /// @dev May revert with ERC20Issuance_Blacklist_NotBlacklistManager
```
removed

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+
+    /// @notice Modifier to check if the caller is a blacklist manager
+    /// @dev May revert with ERC20Issuance_Blacklist_NotBlacklistManager
+    modifier onlyBlacklistManager() {
```
moved

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+        external
+        onlyBlacklistManager
+    {
+        uint totalAccount_ = accounts_.length;
```
done

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    /// @inheritdoc IERC20Issuance_Blacklist_v1
+    function removeFromBlacklistBatchAddresses(address[] calldata accounts_)
```
done

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+    /// @param  from Address tokens are transferred from
+    /// @param  to Address tokens are transferred to
+    /// @param  amount Number of tokens to transfer
+    /// @inheritdoc ERC20Capped
```
okay, thanks

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+    /// @param  to Address tokens are transferred to
+    /// @param  amount Number of tokens to transfer
+    /// @inheritdoc ERC20Capped
+    function _update(address from, address to, uint amount)
```
fixed

vendrell46 commented on src/external/token/ERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,207 @@
+
+    /// @notice Internal function to set a blacklist manager
+    /// @param manager_ Address to set as blacklist manager
+    /// @param allowed_ Whether to grant or revoke blacklist manager role
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_blacklist_v1.sol:
```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.0;
+
+// Imports
```
removed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+            RedemptionState.PROCESSING
+        );
+
+        // Emit event
```
more details added

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        uint projectFeeAmount;
+        uint netDeposit;
+
+        // Get net amount, protocol and project fee amounts. Currently there is no issuance project
```
fixed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        // Burn issued token from user
+        _burn(_msgSender(), _depositAmount);
+
+        // Process the protocol fee. We can re-mint some of the burned tokens, since we aren't paying out
```
fixed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+            _projectFeeCollected(projectFeeAmount);
+        }
+
+        // Revert when the redeem amount is lower than minimum amount the user expects
```
fixed

vendrell46 replied:
now

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+        // Cache Collateral Token
+        IERC20 collateralToken = __Module_orchestrator.fundingManager().token();
+
+        // Require that enough collateral token is held to be redeemable
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    /// @dev    Sets the issuance token.
+    ///         This function overrides the internal function set in {BondingCurveBase_v1}, and
+    ///         it updates the `issuanceToken` state variable and caches the decimals as `_issuanceTokenDecimals`.
```
fixed

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,210 @@
+
+    /// @inheritdoc ILM_ManualExternalPriceSetter_v1
+    /// @dev The price_ parameter should be provided with the same number of decimals as the collateral token.
+    ///      For example, if the collateral token has 6 decimals and the price is 1.5, input should be 1500000.
```
fixed

vendrell46 committed:
fixed natspec

vendrell46 committed:
modifier moved

Zitzak commented on src/external/token/ERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,214 @@
+        string memory name_,
+        string memory symbol_,
+        uint8 decimals_,
+        uint initialSupply_,
```
```suggestion
        uint maxSupply_,
```

vendrell46 replied:
done

vendrell46 replied:
oh shoot, I pushed it to our test branch, I'll push it to this branch right away

vendrell46 replied:
done

Zitzak commented on src/external/token/ERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,214 @@
+    /// @param	name_ Token name.
+    /// @param	symbol_ Token symbol.
+    /// @param	decimals_ Token decimals.
+    /// @param	initialSupply_ Initial token supply.
```
```suggestion
    /// @param	maxSupply_ Max token supply.
```

Zitzak replied:
Looking at the underlying ERC20Issuance I've noticed that the value is actually the max supply, making use of the ERC20Capped instead of the initial supply

vendrell46 replied:
done

vendrell46 replied:
sorry, I pushed it to the wrong branch

Zitzak committed:
feat: update openRedemptionAmount

Zitzak committed:
Merge pull request #4 from ZealynxSecurity/update_open_redemption_amount

Zitzak commented:
@marvinkruse 

In regard to this comment:
https://github.com/InverterNetwork/contracts/pull/696#discussion_r1884002207

The parameter is used as the cap within the ERC20Capped token, which is inherited by the ERC20Issuance and thus ERC20Issuance_Blacklist

marvinkruse commented:
> @marvinkruse 
> 
> 
> 
> In regard to this comment:
> 
> https://github.com/InverterNetwork/contracts/pull/696#discussion_r1884002207
> 
> 
> 
> The parameter is used as the cap within the ERC20Capped token, which is inherited by the ERC20Issuance and thus ERC20Issuance_Blacklist

I was just tagging it as you proposed a change and that change wasn't done yet, I didn't mean to make it seem like I disagreed with anything that was said. Or do I misunderstand?

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
applied suggestions

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 commented on src/external/token/ERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,214 @@
+
+    /// @inheritdoc	IERC20Issuance_Blacklist_v1
+    /// @param	account_ The account to check.
+    /// @return	True if the account is blacklisted.
+    function isBlacklisted(address account_) public view returns (bool) {
+        return _blacklist[account_];
+    }
```
done

vendrell46 commented on src/external/token/ERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,214 @@
+
+    /// @inheritdoc	IERC20Issuance_Blacklist_v1
+    /// @param	account_ The account to check.
+    /// @return	True if the account is a blacklist manager.
+    function isBlacklistManager(address account_) public view returns (bool) {
+        return _isBlacklistManager[account_];
+    }
```
done

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
formatting

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+    error ERC20Issuance_Blacklist_NotBlacklistManager();
+
+    /// @notice Thrown when attempting to mint tokens to a blacklisted address
+    error ERC20Issuance_Blacklist_BlacklistedAddress(address account);
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+    error ERC20Issuance_Blacklist_BlacklistedAddress(address account);
+
+    /// @notice Thrown when batch operation exceeds the maximum allowed size
+    error ERC20Issuance_Blacklist_BatchLimitExceeded(uint provided, uint limit);
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+    //--------------------------------------------------------------------------
+    // External Functions
+
+    /// @notice Checks if an address is blacklisted
+    /// @param account_ The address to check
+    /// @return isBlacklisted_ True if address is blacklisted
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+        view
+        returns (bool isBlacklisted_);
+
+    /// @notice Checks if an address is a blacklist manager
+    /// @param account_ The address to check
+    /// @return isBlacklistManager_ True if address is a blacklist manager
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+        view
+        returns (bool isBlacklistManager_);
+
+    /// @notice Adds an address to blacklist
+    /// @param account_ The address to blacklist
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress
+    function addToBlacklist(address account_) external;
+
+    /// @notice Removes an address from blacklist
+    /// @param account_ The address to remove
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress
+    function removeFromBlacklist(address account_) external;
+
+    /// @notice Adds multiple addresses to blacklist
+    /// @param accounts_ Array of addresses to blacklist
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress
+    ///      The array size should not exceed the block gas limit. Consider using
+    ///      smaller batches (e.g., 100-200 addresses) to ensure transaction success.
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+    function addToBlacklistBatchAddresses(address[] memory accounts_)
+        external;
+
+    /// @notice Removes multiple addresses from blacklist
+    /// @param accounts_ Array of addresses to remove
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress
+    ///      The array size should not exceed the block gas limit. Consider using
+    ///      smaller batches (e.g., 100-200 addresses) to ensure transaction success.
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+    function removeFromBlacklistBatchAddresses(address[] calldata accounts_)
+        external;
+
+    /// @notice Sets or revokes blacklist manager role for an address
+    /// @param manager_ The address to grant or revoke the role from
+    /// @param allowed_ Whether to grant (true) or revoke (false) the role
+    /// @dev May revert with ERC20Issuance_Blacklist_ZeroAddress
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+/**
+ * @title   External Price Oracle Funding Manager with Payment Client
+ *
+ * @notice  A funding manager implementation that uses external oracle price feeds
+ *          for token operations. It integrates payment client functionality and
+ *          supports token redemption through a bonding curve mechanism.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+ *                          our Security Policy at security.inverter.network or
+ *                          email us directly!
+ *
+ * @custom:version   1.0.0
+ *
+ * @custom:standard-version  1.0.0
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    // State Variables
+
+    /// @notice Oracle price feed contract used for price discovery.
+    /// @dev    Contract that provides external price information for token valuation.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    IOraclePrice_v1 private _oracle;
+
+    /// @notice Token that is accepted by this funding manager for deposits.
+    /// @dev    The ERC20 token contract used for collateral in this funding manager.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    IERC20 private _token;
+
+    /// @notice Token decimals of the issuance token.
+    /// @dev    Number of decimal places used by the issuance token for proper decimal handling.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    uint8 private _issuanceTokenDecimals;
+
+    /// @notice Token decimals of the Orchestrator token.
+    /// @dev    Number of decimal places used by the collateral token for proper decimal handling.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    /// @dev    Number of decimal places used by the collateral token for proper decimal handling.
+    uint8 private _collateralTokenDecimals;
+
+    /// @notice Maximum fee that can be charged for sell operations, in basis points.
+    /// @notice Maximum allowed project fee percentage that can be charged when selling tokens.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    /// @notice Maximum allowed project fee percentage that can be charged when selling tokens.
+    uint private _maxProjectSellFee;
+
+    /// @notice Maximum fee that can be charged for buy operations, in basis points.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    /// @dev    Keeps track of the next available order ID to ensure uniqueness.
+    uint private _nextOrderId;
+
+    /// @notice Total amount of collateral tokens currently in redemption process.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    /// @notice Flag indicating if direct operations are only allowed.
+    bool private _isDirectOperationsOnly;
+
+    /// @notice Address of the project treasury which will receive the collateral tokens
```
done

vendrell46 commented on src/external/token/ERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,214 @@
+    IERC20Issuance_Blacklist_v1,
+    ERC20Issuance_v1
+{
+    // --------------------------------------------------------------------------
```
thanks, done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+        Metadata memory metadata_,
+        bytes memory configData_
+    ) external override(Module_v1) initializer {
+        // Initialize parent contracts
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+        // Initialize parent contracts
+        __Module_init(orchestrator_, metadata_);
+
+        // Decode config data
```
done

vendrell46 committed:
formatting

vendrell46 committed:
Merge branch 'dev-1' of https://github.com/ZealynxSecurity/InverterContracts into dev-1

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+        _setOracleAddress(oracle_);
+    }
+
+    /// @notice Gets current buy fee.
+    /// @return buyFee The current buy fee.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+        return _maxBuyFee;
+    }
+
+    /// @notice Gets the maximum project fee that can be charged for sell operations.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+
+    /// @notice Toggles whether the contract only allows direct operations or not.
+    /// @dev    Only the orchestrator admin can call this function.
+    /// @param isDirectOperationsOnly_ The new value for the flag.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    // Internal Functions
+
+    /// @notice Creates and emits a new redemption order.
+    /// @dev    This function wraps the `_createAndEmitOrder` internal function with
+    ///         specified parameters to handle the transaction and direct the proceeds.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    /// @dev    This function wraps the `_sellOrder` internal function with
+    ///         the specified parameters to handle the transaction and direct
+    ///         the proceeds.
+    /// @param  _receiver   The address that will receive the redeemed tokens.
+    /// @param  _depositAmount  The amount of tokens to be sold.
+    /// @param  _minAmountOut   The minimum acceptable amount of proceeds that the
+    ///                         receiver should receive from the sale.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+        return (totalCollateralTokenMovedOut, issuanceFeeAmount);
+    }
+
+    /// @dev    Internal function which only emits the event for amount of project fee
+    ///         collected. The contract does not hold collateral as the payout is managed
+    ///         through a redemption queue.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    }
+
+    /// @dev    Sets the issuance token.
+    ///         This function overrides the internal function set in {BondingCurveBase_v1},
+    ///         and it updates the `issuanceToken` state variable and caches the decimals
+    ///         as `_issuanceTokenDecimals`.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    }
+
+    /// @notice Sets the oracle address
+    /// @dev May revert with Module__FM_PC_ExternalPrice_Redeeming_InvalidOracleInterface
+    /// @param oracleAddress_ The address of the oracle
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+        _oracle = IOraclePrice_v1(oracleAddress_);
+    }
+
+    /// @notice Sets the project treasury address
+    /// @dev May revert with Module__FM_PC_ExternalPrice_Redeeming_InvalidProjectTreasury
+    /// @param projectTreasury_ The address of the project treasury
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    }
+
+    /// @inheritdoc BondingCurveBase_v1
+    /// @dev        Implementation transfer collateral tokens to the project treasury
```
done

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,759 @@
+    }
+
+    /// @inheritdoc RedeemingBondingCurveBase_v1
+    /// @dev    Implementation does not transfer collateral tokens to recipient as the
+    ///         payout is managed through a redemption queue.
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+ * @title   Manual External Price Oracle Implementation
+ *
+ * @notice  This contract provides a manual price feed mechanism for token
+ *          operations, allowing authorized users to set and update prices for
```
fixed

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+ *
+ * @custom:version   1.0.0
+ *
+ * @custom:standard-version  1.0.0
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+    /// @dev    This role should be granted to trusted price feeders only
+    bytes32 public constant PRICE_SETTER_ROLE = "PRICE_SETTER_ROLE";
+
+    /// @notice Number of decimal places used for internal price representation
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+    //--------------------------------------------------------------------------
+    // State Variables
+
+    /// @notice The price for issuing tokens (normalized to INTERNAL_DECIMALS)
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+    /// @notice The price for issuing tokens (normalized to INTERNAL_DECIMALS)
+    uint private _issuancePrice;
+
+    /// @notice The price for redeeming tokens (normalized to INTERNAL_DECIMALS)
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+    /// @notice The price for redeeming tokens (normalized to INTERNAL_DECIMALS)
+    uint private _redemptionPrice;
+
+    /// @notice Decimals of the collateral token (e.g., USDC with 6 decimals)
+    /// @dev    This is the token used to pay/buy with
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+    /// @dev    This is the token used to pay/buy with
+    uint8 private _collateralTokenDecimals;
+
+    /// @notice Decimals of the issuance token (e.g., ISS with 18 decimals)
+    /// @dev    This is the token being bought/sold
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+    /// @dev    This is the token being bought/sold
+    uint8 private _issuanceTokenDecimals;
+
+    /// @dev    Storage gap for upgradeable contracts
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+    /// @dev    Storage gap for upgradeable contracts
+    uint[50] private __gap;
+
+    //--------------------------------------------------------------------------
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+    ) external override(Module_v1) initializer {
+        __Module_init(orchestrator_, metadata_);
+
+        // Decode collateral and issuance token addresses from configData_
```
done

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,215 @@
+
+    /// @notice Normalizes a price from token decimals to internal decimals
+    /// @param  price_ The price to normalize
+    /// @param  tokenDecimals_ The decimals of the token the price is denominated in
```
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,165 @@
+        uint fee, uint maxFee
+    );
+
+    /// @notice	Thrown when the oracle contract does not implement the required interface
```
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,165 @@
+ *
+ * @author  Zealynx Security
+ */
+interface IFM_PC_ExternalPrice_Redeeming_v1 is
```
fixed

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,165 @@
+    error Module__FM_PC_ExternalPrice_Redeeming_InvalidAmount();
+
+    /// @notice	Fee exceeds maximum allowed value
+    /// @param	fee The fee that was attempted to be set
+    /// @param	maxFee The maximum allowed fee
+    error Module__FM_PC_ExternalPrice_Redeeming_FeeExceedsMaximum(
+        uint fee, uint maxFee
+    );
```
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,165 @@
+    /// @param	collateralToken Address of collateral token
+    /// @param	redemptionTime Time of redemption
+    /// @param	state Initial state of the order
+    event RedemptionOrderCreated(
```
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,62 @@
+ *
+ * @author  Zealynx Security
+ */
+interface ILM_ManualExternalPriceSetter_v1 is IOraclePrice_v1 {
```
done

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -0,0 +1,46 @@
+    function getPriceForIssuance() external view returns (uint price_);
+
+    /// @notice	Gets current price for token redemption (selling tokens)
+    /// @return	price_ Current price normalized to issuance token decimals (collateral tokens per 1 issuance token)
```
done

vendrell46 committed:
formatting

vendrell46 commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -0,0 +1,46 @@
+    // External Functions
+
+    /// @notice	Gets current price for token issuance (buying tokens)
+    /// @return	price_ Current price normalized to issuance token decimals (collateral tokens per 1 issuance token)
```
done

Zitzak commented:
Thank you @marvinkruse, then I misunderstood your message. Apologies

scab24 committed:
Modifier buyingIsEnabled and sellingIsEnabled added

marvinkruse commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,223 @@
+        _issuanceTokenDecimals = IERC20Metadata(issuanceToken).decimals();
+    }
+
+    //--------------------------------------------------------------------------
```
Missing space here as well in this contract. Line length may need an adjustment afterwards.

vendrell46 replied:
fixed

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,185 @@
+/**
+ * @title   External Price Funding Manager Interface
+ *
+ * @notice  Interface for a funding manager that uses external price feeds for token
```
Needs to be the same as in the contract, this one is old.

vendrell46 replied:
weird, I remember changing this. Done

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,185 @@
+
+    /// @notice Gets the project treasury address.
+    /// @return address_ The address of the project treasury.
+    function getProjectTreasury() external view returns (address);
```
Same as in the contract, unnamed return value.

vendrell46 replied:
added

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,185 @@
+
+    /// @notice Gets the direct operations only flag.
+    /// @return Whether only direct operations are allowed.
+    function getIsDirectOperationsOnly() external view returns (bool);
```
Same as in the contract, unnamed return value.

vendrell46 replied:
added

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,185 @@
+
+    /// @notice Gets current buy fee.
+    /// @return buyFee The current buy fee.
+    function getBuyFee() external view returns (uint);
```
Same as in the contract, unnamed return value.

vendrell46 replied:
added

vendrell46 replied:
added

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,185 @@
+
+    /// @notice Gets current sell fee.
+    /// @return sellFee The current sell fee.
+    function getSellFee() external view returns (uint);
```
Same as in the contract, unnamed return value.

vendrell46 replied:
added

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,62 @@
+ *                          our Security Policy at security.inverter.network or
+ *                          email us directly!
+ *
+ * @custom:version   v1.0.0
```
```suggestion
 * @custom:version  v1.0.0
```

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,62 @@
+ *
+ * @custom:version   v1.0.0
+ *
+ * @custom:standard-version  v1.0.0
```
```suggestion
 * @custom:standard-version v1.0.0
```

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -0,0 +1,52 @@
+ *                          our Security Policy at security.inverter.network or
+ *                          email us directly!
+ *
+ * @custom:version   v1.0.0
```
```suggestion
 * @custom:version  v1.0.0
```

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @custom:version   v1.0.0
+ *
+ * @custom:standard-version  v1.0.0
```
```suggestion
 * @custom:standard-version v1.0.0
```

vendrell46 committed:
formatting, updating code from parent and adding interface check

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,119 @@
+ * @author  Zealynx Security
+ */
+interface IERC20Issuance_Blacklist_v1 is IERC20Issuance_v1 {
+    //--------------------------------------------------------------------------
```
done

vendrell46 commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,121 @@
+    //--------------------------------------------------------------------------
+    // Events
+
+    /// @notice Emitted when an address is added to the blacklist
```
Apologies for that. I've just gone one by one through all the contracts we've implemented and added dots everywhere there was a comment at the end of the sentences

vendrell46 commented on src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -0,0 +1,223 @@
+    // -------------------------------------------------------------------------
+    // Constants
+
+    /// @notice Role identifier for accounts authorized to set prices
```
done

vendrell46 committed:
formatting

vendrell46 committed:
formatting

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+    uint8 private constant EIGHTEEN_DECIMALS = 18;
+
+    /// @notice Role for whitelisted addresses.
+    bytes32 public constant WHITELIST_ROLE = "WHITELIST_ROLE";
+
+    /// @notice Role for payment queue.
+    bytes32 public constant QUEUE_MANAGER_ROLE = "QUEUE_MANAGER_ROLE";
```
moved

vendrell46 replied:
moved back to contract
![Screenshot 2024-12-16 at 18 00 39](https://github.com/user-attachments/assets/2f3da63c-1f6a-41fa-8e81-66e63eb2774a)

vendrell46 committed:
moved constants

vendrell46 committed:
changed notice to dev

vendrell46 committed:
adding named returned variable

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+        public
+        view
+        override(BondingCurveBase_v1)
+        returns (uint)
```
fixed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+    }
+
+    /// @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    function getProjectTreasury() external view returns (address) {
```
fixed

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+    }
+
+    /// @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    function getIsDirectOperationsOnly() public view returns (bool) {
```
fixed

vendrell46 committed:
modifier renamed and buyingIsEnabled added to buy

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+    // Public Functions
+
+    /// @inheritdoc BondingCurveBase_v1
+    function buy(uint collateralAmount_, uint minAmountOut_)
```
nice catch, we missed that. Added.

marvinkruse replied:
Thank you, Sir!

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+    }
+
+    /// @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    function getSellFee() public view returns (uint) {
```
added

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+    }
+
+    /// @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    function getBuyFee() public view returns (uint) {
```
added

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+    }
+
+    /// @inheritdoc IFM_PC_ExternalPrice_Redeeming_v1
+    function getMaxBuyFee() public view returns (uint) {
```
added

vendrell46 committed:
return named variables to interface

vendrell46 commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,735 @@
+        public
+        view
+        override(ERC20PaymentClientBase_v1, RedeemingBondingCurveBase_v1)
+        returns (bool)
```
added

vendrell46 committed:
formatting

marvinkruse commented on src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol:
```diff
@@ -0,0 +1,123 @@
+import {IERC20Issuance_v1} from "@ex/token/IERC20Issuance_v1.sol";
+
+/**
+ * @title   ERC20 Issuance Token with Blacklist Functionality.
```
Adding another comment as my original one is being ignored (on accident I believe) :D

> Needs to be copied from contract again after the changes to update.

vendrell46 replied:
I would have said I did this, but maybe there was another one. Done

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,742 @@
+        uint projectFeeAmount;
+        uint netDeposit;
+
+        // Get net amount, protocol and project fee amounts. Currently there is no issuance project
```
Adding another comment as my last one got lost I believe:
> The line is too long, needs a break at 80 characters max.

vendrell46 replied:
done

marvinkruse commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -0,0 +1,742 @@
+        // Burn issued token from user.
+        _burn(_msgSender(), _depositAmount);
+
+        // Process the protocol fee. We can re-mint some of the burned tokens, since we aren't paying out
```
Adding another comment as my last one got lost I believe:
> The line is too long, needs a break at 80 characters max.

vendrell46 replied:
done

marvinkruse commented on src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol:
```diff
@@ -0,0 +1,52 @@
+pragma solidity ^0.8.0;
+
+/**
+ * @title   Oracle Price Interface.
```
Natspec section needs to be copied from the contract now that that one was changed. also no dot at the end of a title 🙏

vendrell46 replied:
there isn't a `OraclePrice_v1` implementation, so what do you suggest, @marvinkruse?

This interface is extended in `interface ILM_ManualExternalPriceSetter_v1 is IOraclePrice_v1 `

vendrell46 replied:
dot removed

Zitzak committed:
fix: revert during deployment

Zitzak committed:
Merge pull request #6 from ZealynxSecurity/fix_setting_error

fix: revert during deployment

Zitzak committed:
fix issue with authorization

Zitzak committed:
Merge pull request #7 from ZealynxSecurity/add_internal_setter_function

fix issue with authorization

Zitzak committed:
fix: interface check

Zitzak committed:
Merge pull request #8 from ZealynxSecurity/fix_interface_check

fix: interface check

vendrell46 committed:
interface natspec same as contract

vendrell46 committed:
line too long fixed

vendrell46 committed:
removed dot from title

vendrell46 committed:
_ensureTokenBalance overriden

Zitzak committed:
remove function

Zitzak committed:
Merge branch 'dev-1' into remove_function

Zitzak committed:
fix fmt

Zitzak committed:
Merge pull request #9 from ZealynxSecurity/remove_function

remove function from FM

marvinkruse reviewed (commented):
Good job on getting the changes in, looks great 🙏 Just a simple line being too long, otherwise I couldn't spot anything obvious.

Zitzak committed:
Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

Zitzak committed:
fix: Move LM_ManualPriceSetter into LM folder

Zitzak committed:
Merge pull request #10 from ZealynxSecurity/move_lm_to_right_folder

fix: Move LM_ManualPriceSetter into LM folder

Zitzak reviewed (approved):
Longest PR I've ever seen :D Great work @vendrell46 @scab24
